### PR TITLE
add filter_samples()

### PR DESCRIPTION
**Pull request by vict0rsch**

### CODE REVIEW

**Comment by vict0rsch** [comment]
@51N84D for the sake of the exercise can you write a. test in test_data?

**Comment by vict0rsch** [comment]
@51N84D for the sake of the exercise can you write a test in test_data?

**Comment by 51N84D** [comment]
> @51N84D for the sake of the exercise can you write a test in test_data?

Sure thing

**Comment by 51N84D** [comment]
Also, the `filter_samples()` function uses the instance variable `self.opts` but `self.opts` is never defined. I'll fix that as well

**Comment by 51N84D** [comment]
Oh, and `filter_samples()` gets rid of the `x` variable in sample (since it's technically not a "task" variable) but we probably want to keep that one ;)

**Comment by vict0rsch** [comment]
write the tests, I'll fix that now :)

File: tests/test_data.py [hunk 1]
```diff
@@ -75,9 +84,7 @@
                 [
                     str(
                         {
-                            k: [(s, t.shape) for s, t in v.items()]
-                            if k == "data"
-                            else v
+                            k: [(s, t.shape) for s, t in v.items()] if k == "data" else v
```
Comment (vict0rsch) [inline tests/test_data.py:88]: is that a black artifact?

**Comment by vict0rsch** [comment]
@51N84D  I think this makes more sense:  

```
    self.tasks = set(opts.tasks)
    self.tasks.add("x")

...

    def filter_samples(self):
        """
        Filter out data which is not required for the model's tasks
        as defined in opts.tasks
        """
        self.samples_paths = [
            {k: v for k, v in s.items() if k in self.tasks}
            for s in self.samples_paths
        ]
```

**Comment by vict0rsch** [comment]
samples might not all have the same available data so you can't trust the first sample

**Comment by 51N84D** [comment]
> samples might not all have the same available data so you can't trust the first sample

So better would be to iterate through all the samples and check that the variables are subsets of the task variables?

**Comment by vict0rsch** [comment]
yes, the piece of code I commented above would work don't you agree?

**Comment by 51N84D** [comment]
> yes, the piece of code I commented above would work don't you agree?

I agree that it does what we want in the dataset class, but I meant more how do we verify that it's doing what we want in the test program (which is to check that the sample variables are indeed subsets of the task variables)

File: tests/test_data.py [hunk 2]
```diff
@@ -31,10 +31,11 @@
     # --------------------------------
     # -----  Test task matching  -----
     # --------------------------------
-    ds_vars = set(ds.samples_paths[0].keys())
     tasks = set(opts.tasks)
     tasks.add("x")
-    assert ds_vars.issubset(tasks)
+    for sample_path in ds.samples_paths:
```
Comment (vict0rsch) [inline tests/test_data.py:36]: I really think we should use `filter_samples`! We don't care if some samples have *more* key, we want them to have *at least* the keys in `tasks` and remove extra-ones

Comment (vict0rsch) [reply tests/test_data.py:36]: oups that's in the test sorry!

Comment (vict0rsch) [reply tests/test_data.py:36]: I thought it was in filter_sampels

**Comment by vict0rsch** [comment]
what about now @51N84D ?

File: omnigan/data.py [hunk 3]
```diff
@@ -167,6 +179,7 @@ def check_samples(self):
         exist on the file-system
         """
         for s in self.samples_paths:
+            assert all([t in s for t in self.tasks])
```
Comment (vict0rsch) [inline omnigan/data.py:182]: actually I'm not sure we should have that here 'cause it prevents samples from having missing data. what do you think?  should we enforce all samples having the same data, matching `opts.tasks` ?

**Comment by 51N84D** [comment]
> what about now @51N84D ?

Haha yes, I think we're ready to merge

**Pull request merged**